File: scanner-python/app/services/abuseipdb.py
import httpx
from datetime import datetime, timezone
from app.config import settings
from app.utils.normalize import normalize_abuseipdb, utc_now
from app.models.shared import SourceMetadata

ABUSEIPDB_URL = "https://api.abuseipdb.com/api/v2/check"

async def fetch_abuseipdb(ip: str) -> tuple[dict | None, SourceMetadata]:
    source = SourceMetadata(
        source="abuseipdb",
        retrieved_at=utc_now(),
        success=False
    )
    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(
                ABUSEIPDB_URL,
                headers={
                    "Key": settings.abuseipdb_api_key,
                    "Accept": "application/json"
                },
                params={
                    "ipAddress": ip,
                    "maxAgeInDays": 90
                },
                timeout=10.0
            )
            response.raise_for_status()
            raw = response.json()
            normalized = normalize_abuseipdb(raw)
            source.success = True
            return normalized, source

    except httpx.TimeoutException:
        source.error = "Request timed out"
        return None, source
    except httpx.HTTPStatusError as e:
        source.error = f"HTTP {e.response.status_code}"
        return None, source
    except Exception as e:
        source.error = str(e)
        return None, source

File: scanner-go/internal/services/abuseipdb.go
package services

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/Ahlyx/scanner-go/internal/models"
)

const abuseIPDBURL = "https://api.abuseipdb.com/api/v2/check"

func FetchAbuseIPDB(apiKey, ip string) (*models.AbuseData, models.SourceMetadata) {
	meta := models.SourceMetadata{
		Source:      "abuseipdb",
		RetrievedAt: time.Now().UTC(),
	}

	acquireSem()
	defer releaseSem()

	req, err := http.NewRequest(http.MethodGet, abuseIPDBURL, nil)
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}
	req.Header.Set("Key", apiKey)
	req.Header.Set("Accept", "application/json")
	q := req.URL.Query()
	q.Set("ipAddress", ip)
	q.Set("maxAgeInDays", "90")
	req.URL.RawQuery = q.Encode()

	resp, err := httpClient.Do(req)
	if err != nil {
		errStr := "Request timed out"
		if err != http.ErrHandlerTimeout {
			errStr = err.Error()
		}
		meta.Error = ptr(errStr)
		return nil, meta
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		meta.Error = ptr(fmt.Sprintf("HTTP %d", resp.StatusCode))
		return nil, meta
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}

	var raw struct {
		Data struct {
			AbuseConfidenceScore int     `json:"abuseConfidenceScore"`
			TotalReports         int     `json:"totalReports"`
			LastReportedAt       *string `json:"lastReportedAt"`
			ISP                  *string `json:"isp"`
			UsageType            *string `json:"usageType"`
			IsTor                bool    `json:"isTor"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &raw); err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}

	meta.Success = true
	return &models.AbuseData{
		AbuseScore:   ptr(raw.Data.AbuseConfidenceScore),
		TotalReports: ptr(raw.Data.TotalReports),
		LastReported: raw.Data.LastReportedAt,
		ISP:          raw.Data.ISP,
		UsageType:    raw.Data.UsageType,
		IsTor:        ptr(raw.Data.IsTor),
	}, meta
}
